### DISCUSSION TITLE
how to keep livewire loading state on form submit

### DISCUSSION THREAD

**Discussion by sirwanveisi** [livewire/livewire#4660] [1. Help]
hi,

how can i keep livewire loading state after form submit? (just like attached image)
I want the loading spinner  to continue to be displayed after the form is submitted until the user is redirected to the next page.

---

![Screenshot 2022-02-13 094333](https://user-images.githubusercontent.com/60171454/153741239-a1f63329-5bf8-450b-a218-d569f271ddc9.png)

**Comment by joshhanley** [comment]
@sirwanveisi you could just add a flag on the backend, like `$keepLoading` and set it to true in your request, so you can then display the loading state manually until the redirect happens. Not sure if that will cause a flash or not though, you will need to test it.

Hope this helps!

**Reply by sirwanveisi** [reply]
thanks but i tested this before and the loading stoped on redirecting time!

**Reply by joshhanley** [reply]
@sirwanveisi hmm yeah. One thing you could also try is skipping render on redirect, so in your method that does the redirect, add `$this->skipRender();`

**Reply by sirwanveisi** [reply]
i tried every way you can imagine and the problem still exists :/ wtf

**Reply by joshhanley** [reply]
@sirwanveisi yeah I'm not sure how else to do it, unless you write your own loading handler using javascript and Livewire's javascript object events https://laravel-livewire.com/docs/2.x/lifecycle-hooks#js-hooks

**Reply by ghost** [reply]
@joshhanley Did you find a solution for this?

**Reply by joshhanley** [reply]
@Vincent-Ntriga wasn't me trying to do it, I was just trying to help out 🙂

**Reply by harrisonratcliffe** [reply]
@Vincent-Ntriga @sirwanveisi Did you find any workaround/solution?

**Reply by ghost** [reply]
@harrisonratcliffe  No, I didn't.

**Comment by agustinprod** [comment]
Livewire skips the re-render by default when you redirect. The only option is to set a $this->shouldSkipRender = false; before the redirect and use a custom flag to check if still is loading.

We're having problems too because some people submit forms several times while the redirect is being made, and livewire could manage this easier than devs having to add more code to our apps.

@joshhanley  i think it would be a good idea to keep loading states if the response is a redirect. Or maybe having a flag like `wire:loading.redirects` thats keeps loading states while browsers performs the redirect.

**Reply by joshhanley** [reply]
@killgt yeah that makes sense to me. PR's are welcome! 🙂

**Reply by harrisonratcliffe** [reply]
This would be extremely useful. 

@killgt - have you written a workaround?

**Reply by agustinprod** [reply]
The workaround was setting a variable manually while redirecting and conditionally applying a class, but i don't have time right now to work on a better solution.

**Comment by swen4** [comment]
I stumbled upon this discussion after running into the same issue myself. I have many forms that result in a redirect on success, and was bothered by the disabled attribute and my loading attributes being removed as the page was redirecting.

With the help of Codex CLI I managed to create this drop-in script to prevent this behavior. Use at your own risk, I haven't thoroughly tested this yet, but it seems to work in two different projects of mine so far.

```js
/**
 * Livewire redirect loading helper.
 *
 * Problem: wire:loading elements revert when a Livewire action returns a redirect,
 * allowing double-clicks while the browser navigates.
 *
 * Fix: on redirect effects, re-apply wire:loading toggles (class/attr/display)
 * and disable submit buttons so the loading/disabled state persists until navigation.
 *
 * Usage: import this file into your main bundle so it runs once in the browser.
 * Works with Livewire v3; no project-specific CSS or data attributes required.
 */

const DISPLAY_MAP = {
    inline: "inline",
    block: "block",
    table: "table",
    flex: "flex",
    grid: "grid",
    "inline-flex": "inline-flex",
};

let isInitialized = false;

function initRedirectLoadingPersistence() {
    if (isInitialized) {
        return;
    }

    if (!window.Livewire || typeof window.Livewire.hook !== "function") {
        document.addEventListener("livewire:initialized", () => initRedirectLoadingPersistence(), { once: true });
        return;
    }

    isInitialized = true;

    const redirectingComponents = new WeakSet();

    window.Livewire.hook("effect", ({ component, effects }) => {
        if (!effects || !effects.redirect) {
            return;
        }

        redirectingComponents.add(component);
        applyPersistentLoading(component);
    });

    // Clear any flags when the browser is about to navigate away.
    window.addEventListener("beforeunload", () => {
        redirectingComponents.clear?.();
    });

    function applyPersistentLoading(component) {
        if (!component || !component.el) {
            return;
        }

        const loadingElements = collectLoadingElements(component.el);
        loadingElements.forEach(applyLoadingState);

        disableSubmitButtons(component.el);
    }
}

function collectLoadingElements(root) {
    const nodes = [];
    const elements = [root, ...root.querySelectorAll("*")];

    elements.forEach((el) => {
        const hasLoading = el.getAttributeNames().some((name) => name.startsWith("wire:loading"));

        if (hasLoading) {
            nodes.push(el);
        }
    });

    return nodes;
}

function applyLoadingState(el) {
    el.getAttributeNames()
        .filter((name) => name.startsWith("wire:loading"))
        .forEach((attrName) => {
            const modifiers = attrName.split(".").slice(1);
            const value = el.getAttribute(attrName) || "";

            if (modifiers.includes("class")) {
                const classes = value.split(/\s+/).filter(Boolean);
                if (classes.length === 0) {
                    return;
                }

                if (modifiers.includes("remove")) {
                    classes.forEach((cls) => el.classList.remove(cls));
                } else {
                    el.classList.add(...classes);
                }

                return;
            }

            if (modifiers.includes("attr")) {
                const attrs = value.split(/\s+/).filter(Boolean);
                if (attrs.length === 0) {
                    return;
                }

                attrs.forEach((attr) => {
                    el.setAttribute(attr, "");
                    if (attr === "disabled") {
                        el.setAttribute("aria-disabled", "true");
                    }
                });

                return;
            }

            // Default visibility handling
            if (modifiers.includes("remove")) {
                el.style.setProperty("display", "none", "important");
                return;
            }

            const displayModifier = modifiers.find((mod) => DISPLAY_MAP[mod]);
            const displayValue = DISPLAY_MAP[displayModifier] || "inline-block";
            el.style.setProperty("display", displayValue, "important");
        });
}

function disableSubmitButtons(root) {
    const submits = root.querySelectorAll('button[type="submit"], input[type="submit"]');
    submits.forEach((btn) => {
        btn.setAttribute("disabled", "disabled");
        btn.setAttribute("aria-disabled", "true");
        btn.style.cursor = "not-allowed";
    });
}

// Defer init until Livewire is ready.
if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", initRedirectLoadingPersistence);
} else {
    initRedirectLoadingPersistence();
}
```
